second-analysis-steps/advanced-dirac.md

```diff
@@ -0,0 +1,207 @@
+trf1.backend = Local()
+
+# change the environment  object for each Unit/Master Job
+trf1.unit_splitter = GenericSplitter()
+trf1.unit_splitter.attribute = "application.env"
+trf1.unit_splitter.values = [{'MYENV':'slow_gaussian'}, {'MYENV':'poisson'}, {'MYENV':'flat'}]
+
+#change the argument object for each subjob in the Unit
+trf1.splitter = GenericSplitter()
+trf1.splitter.attribute = 'application.args'
+trf1.splitter.values = [*range(100,1000,100)]
+
+# Append the transform
+t.appendTransform(trf1)
+
+# set the maximum number of active jobs to have running (allows for throttling)
+t.float = 100
+
+# run the Task
+t.run()
+```
+
+There is a lot happening here so we can break it down. `t` is our `CoreTask()` object that will manage all our individual processes. Each distinct process is called a `Transform` and these can be further broken down into `Units` that are generally general configurations of `Transforms` and `subjobs` that are unique.
+
+Hence, we are defining a `Task` to generate three different distributions (gaussian, poisson, flat) over a range of events. Here we take advantage of pythons `*` operator to unpack the `range` into an array. The final thing to note is that we are calling a bash script called `my_script.sh`. We need to define this in order to run python in the appropriate environment. This takes the form.
+
+```
+!/bin/bash
+
+#run the python script that you uploaded
+lb-conda default python distro.py $MYENV $1
+
+``` 
+
+{% callout ".env v .arg" %} 
+
+It is important to note that the subjob splitter and the Unit splitter have to be on separate job parameters (in this case .env and .arg). Otherwise, the last splitter called will overwrite the changes of the first. You will still end up with the correct ammount of jobs but they will be running the wrong parameters! To inspect what parameters your jobs are running with you can use the usual Ganga syntax of
+```
+jobs(job_num) 
+```
+to quicky inspect what was submitted.
+
+{% endcallout %}
+
+Run this the same way you would any Ganga submission script using.
+```
+ganga -i submit.py
+```
+
+You can inspect your Task by looking at `tasks` in the ganga `IPython` window. Do not worry if nothing happens, the `tasks` monitor only refreshes every 30 real seconds! If everything has gone as intended shortly you will have some data files in your local gangadir relating to the generators. How do we add the second transform?
+
+To add the second transform we need to append the following to the Ganga submission script before the run command.
+
+```
+# Create the second transform
+trf2 = CoreTransform()
+trf2.application = Executable()
+trf2.application.exe = File('my_other_script.sh')
+trf2.inputfiles = [LocalFile("compare.py")]
+trf2.outputfiles = [LocalFile("*.png")]
+
+#specify transform dependencies
+task_chain = TaskChainInput()
+task_chain.input_trf_id = trf1.getID()
+trf2.addInputData(task_chain)
+
+trf2.splitter = GangaDatasetSplitter()
```

Comment:
would be useful to explain how this splitter differs from `SplitByFiles` which people are used to see in the davinci-type jobs.

Reply:
Just to be clear on the difference myself before I add that. `SplitByFiles` splits by the file locations in a provided input file/files whereas `GangaDatasetSplitter` splits based on the files inside __GangaInputData.txt__